File: Delimiter_check/queue_delimiter.py
# Program to confirm proper bracket/brace/parenthesis delimiting in a stringLength

# Queue implemented using collections module
from collections import deque

def balanceCheck(userInput):
    counter = 0
    isBalanced = True
    delimQueue = deque()                                    # our queue is technically a deque, but whatever

    for i in userInput:                                     # iterate through user input string
        if userInput[counter] == ('[' or '{' or '('):       # detect opening delimiting character
            delimQueue.append(userInput[counter])           # enqueue opening delimiting character to stack
        elif userInput[counter] == (']' or '}' or ')'):     # detect closing delimiting character
            if not delimQueue:
                isBalanced = False                          # a closing bracket with an empty stack means automatic failure
            # remove matching delimiting characters - there has got to be a more graceful way to do that
            elif ((delimQueue[-1] == '[') and (userInput[counter] == ']')) or (delimQueue[-1] == '{') and (userInput[counter] == '}') or (delimQueue[-1] == '(') and (userInput[counter] == ')'):
                delimQueue.popleft()
        counter += 1

    if delimQueue:                                          # any elements remaining after matching/dequeueing is finished indicates failure
        isBalanced = False
    return isBalanced

# poll user for phrase to check
userInput = input("Enter a string delimited by brackets, parentheses, or braces: ")

if(balanceCheck(userInput)):    # run balance check function
    print("This string is properly delimited.")
else:
    print("This string is NOT properly delimited.")

File: Delimiter_check/queue_delimiter.cpp
// Program to confirm proper bracket/brace/parenthesis delimiting in a string

#include <iostream>
#include <string>
#include <queue>

using namespace std;

bool balanceCheck(string userInput)
{
  bool isBalanced = true;
  int stringLength = userInput.length();
  queue <char> delimQueue;  // queue for storing delimiting characters

  for(int i = 0; i < stringLength; i++) {
      // push opening delimiting characters into the queue
      if(userInput[i] == '[' || userInput[i] == '{' || userInput[i] == '(') delimQueue.push(userInput[i]);

      // detect closing delimiting characters and evaluate balance
      else if(userInput[i] == ']' || userInput[i] == '}' || userInput[i] == ')'){
        if(delimQueue.empty()) isBalanced = false;  // a closing delimiting character with no complement in the queue is an automatic failure

      // remove matching delimiting characters
        else if((delimQueue.front() == '[' && userInput[i] == ']') || (delimQueue.front() == '{' && userInput[i] == '}') || (delimQueue.front() == '(' && userInput[i] == ')')) delimQueue.pop();
      }
  }

  // if there are still elements in the queue after removing all matching delimiting characters, the string is improperly delimited
  if(!(delimQueue.empty())) isBalanced = false;

  return isBalanced;
}

int main()
{
  string userInput;

  // poll user for input
  cout << "Enter a string delimited by brackets, parentheses or braces." << endl;
  cin >> userInput;
  if(balanceCheck(userInput)) cout << "This string is properly delimited." << endl;
  else cout << "This string is NOT properly delimited." << endl;

  return 0;
}
